spring-cloud-deployer-local/src/main/java/org/springframework/cloud/deployer/spi/local/DockerCommandBuilder.java

```diff
@@ -47,47 +54,71 @@
 
 	@Override
 	public String[] buildExecutionCommand(AppDeploymentRequest request, Map<String, String> appInstanceEnv,
-										  Map<String, String> appProperties, Optional<Integer> appInstanceNumber) {
-		List<String> commands = addDockerOptions(request, appInstanceEnv, appProperties, appInstanceNumber);
-		// Add appProperties
-		for (String prop : appProperties.keySet()) {
-			commands.add(String.format("--%s=%s", prop, appProperties.get(prop)));
-		}
+										  Optional<Integer> appInstanceNumber) {
+		List<String> commands = addDockerOptions(request, appInstanceEnv, appInstanceNumber);
 		commands.addAll(request.getCommandlineArguments());
 		logger.debug("Docker Command = " + commands);
 		return commands.toArray(new String[0]);
 	}
 
 	private List<String> addDockerOptions(AppDeploymentRequest request, Map<String, String> appInstanceEnv,
-										  Map<String, String> appProperties, Optional<Integer> appInstanceNumber) {
+										  Optional<Integer> appInstanceNumber) {
 		List<String> commands = new ArrayList<>();
 		commands.add("docker");
 		commands.add("run");
+
 		// Add env vars
 		for (String env : appInstanceEnv.keySet()) {
 			commands.add("-e");
 			commands.add(String.format("%s=%s", env, appInstanceEnv.get(env)));
 		}
-		if (appProperties.containsKey(LocalAppDeployer.SERVER_PORT_KEY)) {
-			String port = appProperties.get(LocalAppDeployer.SERVER_PORT_KEY);
-			commands.add("-p");
-			commands.add(String.format("%s:%s", port, port));
-		}
+
+		setPort(commands, appInstanceEnv);
+
 		if(request.getDeploymentProperties().containsKey(DOCKER_CONTAINER_NAME_KEY)) {
 			if(appInstanceNumber.isPresent()) {
 				commands.add(String.format("--name=%s-%d", request.getDeploymentProperties().get(DOCKER_CONTAINER_NAME_KEY), appInstanceNumber.get()));
 			} else {
 				commands.add(String.format("--name=%s", request.getDeploymentProperties().get(DOCKER_CONTAINER_NAME_KEY)));
 			}
 		}
+
 		DockerResource dockerResource = (DockerResource) request.getResource();
+
 		try {
 			String dockerImageURI = dockerResource.getURI().toString();
 			commands.add(dockerImageURI.substring("docker:".length()));
 		}
 		catch (IOException e) {
 			throw new IllegalStateException(e);
 		}
+
 		return commands;
 	}
+
+	private void setPort(List<String> commands, Map<String, String> appInstanceEnv) {
+
+		String port;
+
+		if(appInstanceEnv.containsKey(AbstractLocalDeployerSupport.SPRING_APPLICATION_JSON)) {
+			Map<String, String> properties = new HashMap<>();
+
+			try {
+				properties.putAll(OBJECT_MAPPER.readValue(appInstanceEnv.get(AbstractLocalDeployerSupport.SPRING_APPLICATION_JSON), new TypeReference<HashMap<String,Object>>() {}));
+			}
+			catch (IOException e) {
+				throw new IllegalArgumentException("Unable to determine server port from SPRING_APPLICATION_JSON");
+			}
+
+			port = properties.get(LocalAppDeployer.SERVER_PORT_KEY);
```

Comment:
Is there a possibility that port is not provided? Null check? 
I see below you're doing `hasText`, but is it event valid if it hasn't?

Reply:
I think it is valid.  Since this could also be deploying a task which may not have an embedded server.